# ![indrop](https://github.com/CRG-CNAG/BioCoreMiscOpen/blob/master/logo/biocore-logo_small.png) indrop

Indrops analysis pipeline at BioCore@CRG

The pipeline is based on the DropEST tool:
https://github.com/hms-dbmi/dropEst

The current supported version is v1 & v2

 File 1: barcode reads. Structure:
 Cell barcode, part 1
 Spacer
 Cell barcode, part 2
 UMI
 
 
 File 2: gene reads


## The pipeline
1. QC: Run FastQC on raw reads. It stores the results within **QC** folder.
1. Indexing: It makes the index of the genome by using STAR.
1. dropTag: It tags the reads by read
1. Merging overlapping pairs: it uses flash to merge overlapping paired ends (optional).
1. Assembling: It assembles trimmed (eventually merged) reads by using SPAdes assembler. Results are stored within the filder **Assembly**
1. Alignment: It aligns assemlbe scaffolds to the feature DB by using blast. Results in tabular format are stored within the folder **Blast**
1. The scaffolds are also scanned for the presence of RE sites using Emboss' restrict tool and list of common enzymes specified in **params.config**. Results are stored in **REsites** folder.
1. The blast output and the RE sites are used for generating a plot using Circular Genome Viewer (http://wishart.biology.ualberta.ca/cgview/) and a genBank file for each sample.
1. Finally a report is generated by using MultiQC and sent by mail.


